In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split


In [ ]:
url = "https://raw.githubusercontent.com/JWarmenhoven/ISLR-python/master/Notebooks/Data/Advertising.csv"

df = pd.read_csv(url)

print(df.head())

   Unnamed: 0     TV  Radio  Newspaper  Sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9


In [ ]:
#Split Data
train, test = train_test_split(df, test_size=0.3, random_state=1)

print(train.shape)
print(test.shape)
# training set to fit all your models and the test set to evaluate them.

(140, 5)
(60, 5)


In [ ]:
#Fit & Interpret: Fit a simple LinearRegression using only the three original features (TV, Radio, Newspaper).

In [ ]:

X = df[["TV", "Radio", "Newspaper"]]  # features
y = df["Sales"]   # dependent variable

In [ ]:
linear=LinearRegression()
linear.fit(X,y)
linear_test_prediction=linear.predict(X)
linear_train_prediction=linear.predict(X)

In [ ]:
# Extract features (X) and target (y) from the train and test df
X_train = train[["TV", "Radio", "Newspaper"]]
y_train = train["Sales"]

X_test = test[["TV", "Radio", "Newspaper"]]
y_test = test["Sales"]

print("X_train shape:", X_train.shape)
print("y_train shape:", y_train.shape)
print("X_test shape:", X_test.shape)
print("y_test shape:", y_test.shape)

X_train shape: (140, 3)
y_train shape: (140,)
X_test shape: (60, 3)
y_test shape: (60,)


In [ ]:
linear=LinearRegression()
linear.fit(X_train,y_train)
linear_test_prediction=linear.predict(X_test)
linear_train_prediction=linear.predict(X_train)

In [ ]:
from math import sqrt
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error

# TEST
linear_r2_test = r2_score(y_test, linear_test_prediction)
linear_rmse_test = sqrt(mean_squared_error(y_test, linear_test_prediction))
linear_mae_test = mean_absolute_error(y_test, linear_test_prediction)

# TRAIN
linear_r2_train = r2_score(y_train, linear_train_prediction)
linear_rmse_train = sqrt(mean_squared_error(y_train, linear_train_prediction))
linear_mae_train = mean_absolute_error(y_train, linear_train_prediction)
# summary
r2_df = pd.DataFrame({
    'Model': ['Linear'],
    'R-squared train': [linear_r2_train],
    'R-squared test': [linear_r2_test],
    'RMSE train': [linear_rmse_train],
    'RMSE test': [linear_rmse_test],
    'MAE train': [linear_mae_train],
    'MAE test': [linear_mae_test]
})

r2_df

#Write Down Coefficients and Performance metrics: Note the coefficients for TV, Radio, and Newspaper and its performance metrics on both train set and test set. This will be your baseline model.
# high R2-RMSE Train> RMSE Test so.Same with MAE. Model better in test than in train


,Model,R-squared train,R-squared test,RMSE train,RMSE test,MAE train,MAE test
0,Linear,0.885005,0.922461,1.789726,1.388857,1.374654,1.054833


In [ ]:
#PART 2
#Create Polynomial Features:
#Make sure to fit_transform on your training data and only .transform your test data.
poly_transform=PolynomialFeatures(degree=5)
# fit transform for train
X_train_poly = poly_transform.fit_transform(X_train)

## Transform X_test separately

X_test_poly = poly_transform.transform(X_test)

print("X_train original:", X_train.shape)
print("X_train_poly:", X_train_poly.shape)



X_train original: (140, 3)
X_train_poly: (140, 56)


In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

#only train
X_train_poly_scaled = scaler.fit_transform(X_train_poly)

# for test, only use transform
X_test_poly_scaled = scaler.transform(X_test_poly)

print("X_train_poly_scaled:", X_train_poly_scaled.shape)
print("X_test_poly_scaled:", X_test_poly_scaled.shape)

X_train_poly_scaled: (140, 56)
X_test_poly_scaled: (60, 56)


In [ ]:
linear_poly=LinearRegression()
linear_poly.fit(X_train_poly_scaled,y_train)

linear_poly_train_prediction=linear_poly.predict(X_train_poly_scaled)
linear_poly_test_prediction=linear_poly.predict(X_test_poly_scaled)





In [ ]:
from math import sqrt
linear_poly_r2_test = r2_score(y_test, linear_poly_test_prediction) # TEST
linear_poly_rmse_test = sqrt(mean_squared_error(y_test, linear_poly_test_prediction))
linear_poly_mae_test = mean_absolute_error(y_test, linear_poly_test_prediction)

linear_poly_r2_train = r2_score(y_train, linear_poly_train_prediction) #TRAIN
linear_poly_rmse_train = sqrt(mean_squared_error(y_train, linear_poly_train_prediction))
linear_poly_mae_train = mean_absolute_error(y_train, linear_poly_train_prediction)

r2_df = pd.DataFrame({
    'Model': ['Linear','Linear Poly'],
    'R-squared train': [linear_r2_train,linear_poly_r2_train],
    'R-squared test': [linear_r2_test,linear_poly_r2_test],
    'RMSE train': [linear_rmse_train,linear_poly_rmse_train],
    'RMSE test': [linear_rmse_test,linear_poly_rmse_test],
    'MAE train': [linear_mae_train,linear_poly_mae_train],
    'MAE test': [linear_mae_test,linear_poly_mae_test]
})
r2_df



,Model,R-squared train,R-squared test,RMSE train,RMSE test,MAE train,MAE test
0,Linear,0.885005,0.922461,1.789726,1.388857,1.374654,1.054833
1,Linear Poly,0.997769,0.794384,0.249258,2.261647,0.193958,0.736791


In [ ]:
# coefs
print("Intercept:", linear_poly.intercept_)
print("Coef(model.coef_):")
print(linear_poly.coef_)


Intercept: 13.79142857142856
Coef(model.coef_):
[ 3.68552982e-10  1.22742490e+01 -7.76103224e+00  9.01866144e-01
 -4.31575466e+01  2.79825459e+01 -1.01377760e+01  3.09398521e+01
  2.09594948e+00  1.28001448e+00  7.39531236e+01 -3.56924222e+01
  2.94761695e+01 -4.35427624e+01 -1.17664289e+01  1.81038598e+01
 -3.97053040e+01 -1.65686122e+01  4.95586323e+00 -1.13892678e+01
 -5.76905939e+01  3.06062106e+01 -4.57347515e+01  1.29549692e+01
  3.88652340e+01 -2.45070530e+01  4.13034760e+01 -3.77617140e+00
 -1.06280368e+00 -1.12261241e+01  2.00708382e+01  2.49468454e+01
 -7.27929523e+00  5.22590984e+00  1.47522618e+01  1.64650766e+01
 -1.33226729e+01  2.37645678e+01  1.18704086e+01 -2.65118227e+01
  1.38147435e+01 -2.60274926e+01  1.20134411e+01 -1.09431937e+01
  7.90664647e+00  6.54909110e-01 -1.13031454e+01  8.82546190e+00
 -3.02462517e-01  1.88347147e+00 -6.99295808e+00 -1.19952673e+00
 -1.52045298e+01  2.02355821e+01 -1.63455393e+01 -3.15980337e+00]


# Questions
What do you observe about the coefficients? Are they large or small? R/ Coef are diverse with no clear pattern related to the signs.Some of them are large and others close to zero.

Do they make any intuitive sense? What does this tell you about the risk of this model? (Hint: They will likely be huge and non-sensical, a classic sign of overfitting). R/ Hard to develop a clear interpretation. We see a sign of overlifting due the non-intuitive coefficients because model is adding to it the fitted noise in the training data.

Print the metrics for this complex model on the training set.R/
On the training set, the complex polynomial model achieves an almost perfect R2= 0.998 and very small errors, RMSE= 0.25, MAE = 0.19. This means it fits the training data extremely well.

Print the R-squared score for this same model on the test set.R/
 R2 test= 0.794384

What do you observe? What does the difference between these two scores (and the baseline score from Part 1) tell you about this model? Is this a good model? R/ The polynomial model has an almost perfect R2 for the train set but for the test we have 0.79. In the simple linear we have 0.92. This difference show that model perfom worse than the first in the test set, being a clear sign of overfitting. It memorize training data instead of develop or learning a generalizable relationship. In conclusiion, is not a good model and the first on eis even better.






In [ ]:
from sklearn.linear_model import Ridge

# PART 3
#Fit Ridge: Fit a Ridge model on the same scaled, polynomial training data.

ridge = Ridge(alpha=0.005)
ridge.fit(X_train_poly_scaled, y_train)
ridge_train_pred = ridge.predict(X_train_poly_scaled)
ridge_test_pred = ridge.predict(X_test_poly_scaled)

# metrics
ridge_r2_train = r2_score(y_train, ridge_train_pred)
ridge_r2_test = r2_score(y_test, ridge_test_pred)
ridge_rmse_train = sqrt(mean_squared_error(y_train, ridge_train_pred))
ridge_rmse_test = sqrt(mean_squared_error(y_test, ridge_test_pred))
ridge_mae_train = mean_absolute_error(y_train, ridge_train_pred)
ridge_mae_test = mean_absolute_error(y_test, ridge_test_pred)

print("Ridge (poly, scaled) metrics:")
print("R^2 train:", ridge_r2_train)
print("R^2 test :", ridge_r2_test)
print("RMSE train:", ridge_rmse_train)
print("RMSE test :", ridge_rmse_test)
print("MAE train:", ridge_mae_train)
print("MAE test :", ridge_mae_test)

Ridge (poly, scaled) metrics:
R^2 train: 0.9952606274930826
R^2 test : 0.9839160462847725
RMSE train: 0.3633357150374008
RMSE test : 0.632546952178218
MAE train: 0.26947635838025685
MAE test : 0.37559905371389013


In [ ]:
from sklearn.linear_model import Lasso

#Fit Lasso: Fit a Lasso model on the same scaled, polynomial training data.
lasso = Lasso(alpha=0.0001)  #
lasso.fit(X_train_poly_scaled, y_train)

# Pred
lasso_train_pred = lasso.predict(X_train_poly_scaled)
lasso_test_pred = lasso.predict(X_test_poly_scaled)

# Metrics
lasso_r2_train = r2_score(y_train, lasso_train_pred)
lasso_r2_test = r2_score(y_test, lasso_test_pred)
lasso_rmse_train = sqrt(mean_squared_error(y_train, lasso_train_pred))
lasso_rmse_test = sqrt(mean_squared_error(y_test, lasso_test_pred))
lasso_mae_train = mean_absolute_error(y_train, lasso_train_pred)
lasso_mae_test = mean_absolute_error(y_test, lasso_test_pred)

print("Lasso (poly, scaled) metrics:")
print("R^2 train:", lasso_r2_train)
print("R^2 test :", lasso_r2_test)
print("RMSE train:", lasso_rmse_train)
print("RMSE test :", lasso_rmse_test)
print("MAE train:", lasso_mae_train)
print("MAE test :", lasso_mae_test)

Lasso (poly, scaled) metrics:
R^2 train: 0.9913271544233577
R^2 test : 0.9924786129262542
RMSE train: 0.49150560466275345
RMSE test : 0.4325589773946572
MAE train: 0.3307577503809882
MAE test : 0.3266716687953282


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.577e+01, tolerance: 3.900e-01
  model = cd_fast.enet_coordinate_descent(


In [ ]:
results_df = pd.DataFrame({
    'Model': ['Linear', 'Linear Poly', 'Ridge Poly', 'Lasso Poly'],
    'R-squared train': [linear_r2_train, linear_poly_r2_train, ridge_r2_train, lasso_r2_train],
    'R-squared test':  [linear_r2_test,  linear_poly_r2_test,  ridge_r2_test,  lasso_r2_test],
    'RMSE train':      [linear_rmse_train, linear_poly_rmse_train, ridge_rmse_train, lasso_rmse_train],
    'RMSE test':       [linear_rmse_test,  linear_poly_rmse_test,  ridge_rmse_test,  lasso_rmse_test],
    'MAE train':       [linear_mae_train,  linear_poly_mae_train,  ridge_mae_train,  lasso_mae_train],
    'MAE test':        [linear_mae_test,   linear_poly_mae_test,   ridge_mae_test,   lasso_mae_test]
})
results_df


,Model,R-squared train,R-squared test,RMSE train,RMSE test,MAE train,MAE test
0,Linear,0.885005,0.922461,1.789726,1.388857,1.374654,1.054833
1,Linear Poly,0.997769,0.794384,0.249258,2.261647,0.193958,0.736791
2,Ridge Poly,0.995261,0.983916,0.363336,0.632547,0.269476,0.375599
3,Lasso Poly,0.991327,0.992479,0.491506,0.432559,0.330758,0.326672


In [ ]:
coefficients = lasso.coef_
coef_df = pd.DataFrame({
    "feature": poly_transform.get_feature_names_out(X_train.columns),
    "coef": coefficients
}).fillna(0)

coef_df



,feature,coef
0,1,0.000000
1,TV,6.528809
2,Radio,0.724569
3,Newspaper,0.622863
4,TV^2,-7.259272
5,TV Radio,4.886524
6,TV Newspaper,-1.607599
7,Radio^2,-0.891445
8,Radio Newspaper,-0.445696
9,Newspaper^2,-0.270565


Question 2 (Analysis & Performance):
What do you observe about the coefficients from the two new models now? Are they still large or small? Make some comments about the changes.R/ Ridge coefs has moderate size and more stability. Lasso set some coeff to zero. This change allow a more accurate interpretation


Look at the coefficients from your Lasso model. How many features did it set to zero? What does this tell you about the 'true' drivers of sales?"R/ Lasso set only a few features to exactly zero, and others keep being small. This indicates than only some polynomial terms are importatnt to explain sales. For exapmple, pure weight of radio or newspapers are not that relevant.


What is the performance metrics for your Ridge model on the train set and test set? R/R2 train=0.995261?-RMSE (train)= 0.363336- MAE (train)= 0.26947 R2 (test)=0.983916-RMSE (test)=0.632547-MAE (test)=0.375599

What is the performance metrics for your Lasso model on the train set and test set? R/R2 (train)= 0.991327-RMSE (train)=0.491506-MAE(train)=0.330758, R2 (test)=0.992479-RMSE (test)=0.432559-MAE (test)=0.326672

How do these scores compare to the 'overfit' model's test score? What does this prove about the value of regularization? R/The test performance is really enhace by using these 2 models having a higher R2 and lower testing error. Regularization helps to penalize large coefficients and control in a good way the model complexity.These 2 models are good to prevent overfitting and develop models that we can generalize in a better way.


Question 3 (The Verdict):
In the end, after trying a simple model, an overfit complex model, and two regularized models, what is your final recommendation to the CMO? Which channels (TV, Radio, Newspaper) are the most reliable drivers of sales? R/  Best choice would be a Lasso model that achieve a large R2 and smaller errors. Thanks to the regularization, it helps avoidding the overfitting and produce better coeffs.
According feautures, TV is the biggest driver of sales, with Radio as the second most important while newspaper is the least reliable driver. I would recommend TV as the element most key to prioritize and using radio as a complementary tool.